host/helpers.go

```diff
@@ -9,3 +16,54 @@ func InfoFromHost(h Host) *peer.AddrInfo {
 		Addrs: h.Addrs(),
 	}
 }
+
+// SignedRoutingStateFromHost returns a SignedRoutingState record containing
+// the Host's listen addresses, signed with the Host's private key.
+//
+// By default, only publicly routable addresses will be included.
+// To include loopback and LAN addresses, pass in the IncludeLocalAddrs option:
+//
+//    state := SignedRoutingStateFromHost(h, IncludeLocalAddrs)
+func SignedRoutingStateFromHost(h minimalHost, opts ...Option) (*routing.SignedRoutingState, error) {
+	cfg := config{}
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+
+	privKey := h.Peerstore().PrivKey(h.ID())
+	if privKey == nil {
+		return nil, errors.New("unable to find host's private key in peerstore")
+	}
+
+	var addrs []ma.Multiaddr
+	if cfg.includeLocalAddrs {
+		addrs = h.Addrs()
+	} else {
+		for _, a := range h.Addrs() {
+			if manet.IsPublicAddr(a) {
+				addrs = append(addrs, a)
+			}
+		}
+	}
+
+	return routing.MakeSignedRoutingState(privKey, addrs)
+}
+
+// IncludeLocalAddrs can be passed into SignedRoutingStateFromHost to
+// produce a routing record with LAN and loopback addresses included.
+func IncludeLocalAddrs(cfg *config) {
+	cfg.includeLocalAddrs = true
+}
+
+// minimalHost is the subset of the Host interface that's required by
+// SignedRoutingStateFromHost.
+type minimalHost interface {
```

Comment:
I added this to make testing a bit simpler, but can remove it if it makes things more confusing. I still ended up having to define a ton of stubs to mock out the `Peerstore`, so it doesn't really help that much 😛